### PR TITLE
Catlearn radolm graph samplers

### PR DESCRIPTION
**Pull request by Previsou**
Tests are still failing. I suspect some issue remaining in DirectedGraph pruning and subgraph - or perhaps in my own code.

### CODE REVIEW

**Comment by Previsou** [comment]
Alright, now working. The subsampling tests could probably use an expansion though

**Comment by Kuraao** [comment]
> Alright, now working. The subsampling tests could probably use an expansion though

Nice. The thing you fixed had been lurking around since like, the beginning of this graph_utils file. Finally fixed :D

I didn't take a look at the rest yet, in progress.

File: catlearn/graph_utils.py [hunk 1]
```diff
@@ -536,3 +540,129 @@ def generate_random_graph(
...
+    - ranking: probability distribution over the input graph vertices
+    - rng: Random generator
+
+    Returns:
+    A valid random subgraph
+
+    """
+    ranks = list(ranking(igraph).items())
+    vertices = [v for v, _ in ranks]
+    weights = [w for _, w in ranks]
```
Comment (Kuraao) [inline catlearn/graph_utils.py:567]: vertices, weights = zip(*ranks)

Don't underestimate penguins

File: catlearn/graph_utils.py [hunk 2]
```diff
@@ -536,3 +540,129 @@ def generate_random_graph(
     # return a randomly chosen graph in the factory's memory
     idx = random_generator.randint(0, factory.nb_graphs - 1)
     return factory.graphs[idx]
+
+
+def sample(
+        igraph: DirectedGraph[NodeType],
```
Comment (Kuraao) [inline catlearn/graph_utils.py:546]: There is only one graph around here, why not simply calling it "graph"?
If not I would prefer input_graph to igraph

File: catlearn/graph_utils.py [hunk 3]
```diff
@@ -536,3 +540,129 @@ def generate_random_graph(
...
+    """
+    ranks = list(ranking(igraph).items())
+    vertices = [v for v, _ in ranks]
+    weights = [w for _, w in ranks]
+    sampled_vertices = set(rng.choices(vertices, weights, k=int(sample_vertices_size)))
+    return igraph.subgraph(sampled_vertices)
+
+
+def pagerank_sample(
+        igraph: DirectedGraph[NodeType],
```
Comment (Kuraao) [inline catlearn/graph_utils.py:573]: same comment again on the argument name. Assume the same comment for all other functions.

File: catlearn/graph_utils.py [hunk 4]
```diff
@@ -536,3 +540,129 @@ def generate_random_graph(
...
+    """
+    Sample a random subgraph of `igraph` with a uniform probability over vertices
+
+    Details of parameters: see `sample`
+    """
+    n = len(igraph)
+    return sample(igraph, sample_vertices_size,
+                  lambda G: {v: 1.0/n for v in G}, rng)
+
+def random_walk_sample(
```
Comment (Kuraao) [inline catlearn/graph_utils.py:633]: Open question: Is the current function really what we want?
The way I see it there is 2 options:
- the current one, where we keep only the nodes which have been sampled
- the one in which instead we keep the edge which have been used during the walk (dunno if that's even practical using networkx). (obviously we also only keep the visited nodes)

File: tests/test_graph_utils.py [hunk 5]
```diff
@@ -507,3 +509,103 @@ def test_result_equality(
...
+        factory = GraphRandomFactory(
+            [0.15, 0.15, 0.15, 0.15, 0.14],
+            10, 0.15, rng)
+        # go through generation steps
+        for _ in range(5):
+            next(factory)  # type: ignore
+        return factory.graphs[0]
+
+    def test_sample(self):
+        """ Test sample respect given probability distribution """
```
Comment (Kuraao) [inline tests/test_graph_utils.py:530]: Did you forget that one? :p

File: tests/test_graph_utils.py [hunk 6]
```diff
@@ -507,3 +509,103 @@ def test_result_equality(
...
+
+
+class TestSubgraphSampling:
+
+    @staticmethod
+    def get_graph(rng):
+        """
+        Randomly create somewhat complex graphs, then return a complex one
+        """
+        factory = GraphRandomFactory(
```
Comment (Kuraao) [inline tests/test_graph_utils.py:521]: I was thinking the arguments to GraphRandomFactory should be test parameters.
Maybe use a fixture, or at least define it somewhere else, outside the method.

File: tests/test_graph_utils.py [hunk 7]
```diff
@@ -507,3 +509,103 @@ def test_result_equality(
...
+            [0.15, 0.15, 0.15, 0.15, 0.14],
+            10, 0.15, rng)
+        # go through generation steps
+        for _ in range(5):
```
Comment (Kuraao) [inline tests/test_graph_utils.py:525]: Same as above for the number of iterations

File: tests/test_graph_utils.py [hunk 8]
```diff
@@ -507,3 +509,103 @@ def test_result_equality(
...
+
+    def test_sample(self):
+        """ Test sample respect given probability distribution """
+        pass
+
+    def test_uniform(self):
+        """ Sanity checks on uniform sampler """
+        rng = random.Random()
+        g = self.get_graph(rng)
+        n_vertices = max(1, len(g) - 4)
```
Comment (Kuraao) [inline tests/test_graph_utils.py:537]: why the -4?

File: tests/test_graph_utils.py [hunk 9]
```diff
@@ -507,3 +509,103 @@ def test_result_equality(
...
+        sg = uniform_sample(g, n_vertices, rng)
+        assert 1 <= len(sg) <= n_vertices
+        assert all(v in g for v in sg)
+
+    # Sometimes the RandomFactory fails (erode) or the algorithm does not converge
```
Comment (Kuraao) [inline tests/test_graph_utils.py:542]: So this still happens? (RandomFactory failure). I thought your fix would have made this go away, too bad :'(

File: tests/test_graph_utils.py [hunk 10]
```diff
@@ -507,3 +509,103 @@ def test_result_equality(
...
+        rng = random.Random()
+        g = self.get_graph(rng)
+        seed = rng.choice(list(g))
+        sg = random_walk_sample(g, rng, 12, seeds=[seed])
+        assert 1 <= len(sg) <= 12
+        assert len(sg.over(seed)) <= 1 # empty, or self-reference
+
+    # Sometimes the RandomFactory fails (erode)
+    @pytest.mark.flaky(reruns=3)
+    def test_random_walk_multiple_roots(self):
```
Comment (Kuraao) [inline tests/test_graph_utils.py:594]: I had a failure when running the test (despite the reruns).
Maybe completely normal, but could you describe what's supposed to happen when it fails? (in relation to the comment).

In my case I got a failure on the assert:
assert 1 <= sum(1 for v in sg if len(sg.over(v)) <= 1) <= n_seeds

(the sum is equal to 5 and n_seed to 3)

**Review by Kuraao — CHANGES_REQUESTED** [review]
Apart from the comments on things to modify/open questions:

- I have a problem with TestDirectedGraph.test_stringify, which I find quite strange. It fails on 2 cases:
graph = DirectedGraph(0.5 > {'a': {}}, a > {}, 2 > {0.5: {}, 'a': {}})
expected_graph = DirectedGraph(0 > {2: {}}, 2 > {}, 1 > {0: {}, 2: {}})

graph = DirectedGraph(0.5 > {'a': {}}, a > {}, 2 > {0.5: {}, 'a': {}})
expected_graph = DirectedGraph(0x0 > {'0x2': {}}, 0x2 > {}, 0x1 > {'0x0': {}, '0x2': {}})

- I don't know what radolm means :p But seriously, since the name of the branch appears in the commit messages these should be fixed.

File: tests/test_graph_utils.py (hunk 8 above)
Comment (Previsou) [reply tests/test_graph_utils.py:537]: Why not? Arbitrary number

File: tests/test_graph_utils.py (hunk 9 above)
Comment (Previsou) [reply tests/test_graph_utils.py:542]: Nope, leftover comment. To be removed

File: tests/test_graph_utils.py (hunk 10 above)
Comment (Previsou) [reply tests/test_graph_utils.py:594]: I'll look into it. As per the other non-related failures, I have the same issue with python 3.8 - and not with 3.7 :/

File: tests/test_graph_utils.py (hunk 8 above)
Comment (Previsou) [reply tests/test_graph_utils.py:537]: Arbitrary number

**Comment by Previsou** [comment]
Existing tests are failing because of new Pickle protocol 5, introduced in Python 3.8  The ordering changes

**Comment by Previsou** [comment]
@Kuraao I'll create another pull request, since I had to change the branch name.

File: tests/test_graph_utils.py (hunk 5 above)
Comment (Previsou) [reply tests/test_graph_utils.py:530]: Yup, Added it (see new pull request)

File: tests/test_graph_utils.py (hunk 10 above)
Comment (Previsou) [reply tests/test_graph_utils.py:594]: Was indeed flaky. I was counting vertices which were the start of the sampled paths. Using `over` method, I thought to capture empty or self-ereference with the `len` only, which is wrong.

Setting up a stricter test solved the issue

**Comment by Previsou** [comment]
Moving to https://github.com/arreason/CategoryLearning/pull/24

**Pull request closed without merging**